src/mscorlib/src/System/Text/StringBuilder.cs

```diff
@@ -1362,57 +1363,73 @@ void ISerializable.GetObjectData(SerializationInfo info, StreamingContext contex
 
                 while (pos < len && (ch = format[pos]) == ' ') pos++;
                 Object arg = args[index];
-                StringBuilder fmt = null;
+                String itemFormat = null;
                 if (ch == ':') {
                     pos++;
+                    int startPos = pos;
+
                     while (true) {
                         if (pos == len) FormatError();
                         ch = format[pos];
                         pos++;
-                        if (ch == '{')
-                        {
-                            if (pos < len && format[pos] == '{')  // Treat as escape character for {{
-                                pos++;
-                            else
-                                FormatError();
-                        }
-                        else if (ch == '}')
+
+                        if (ch == '}' || ch == '{')
                         {
-                            if (pos < len && format[pos] == '}')  // Treat as escape character for }}
-                                pos++;
+                            if (ch == '{')
+                            {
+                                if (pos < len && format[pos] == '{')  // Treat as escape character for {{
+                                    pos++;
+                                else
+                                    FormatError();
+                            }
                             else
                             {
-                                pos--;
-                                break;
+                                if (pos < len && format[pos] == '}')  // Treat as escape character for }}
+                                    pos++;
+                                else
+                                {
+                                    pos--;
+                                    break;
+                                }
                             }
+
+                            // Reaching here means the curly bracket has been escaped
+                            // so we need to build up the format string in segments
+                            if (unescapedItemFormat == null)
+                            {
+                                unescapedItemFormat = new StringBuilder();
+                            }
+                            unescapedItemFormat.Append(format, startPos, pos - startPos - 1);
+                            startPos = pos;
                         }
+                    }
 
-                        if (fmt == null) {
-                            fmt = new StringBuilder();
+                    if (unescapedItemFormat == null || unescapedItemFormat.Length == 0)
```

Comment:
If escaping is needed the length will never be zero. This is to protect against `"{0:AAA{{BBB} {1:ZZZ}"` where the temp StringBuilder is created and cleared for {0} but still exists for {1}.